presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/TableMetadataPageSource.java

```diff
@@ -0,0 +1,410 @@
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        checkState(!closed, "TableMetadataPageSource is closed");
+        assureLoaded();
+        return !pageIterator.hasNext();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        checkState(!closed, "TableMetadataPageSource is closed");
+        assureLoaded();
+        return pageIterator.next();
+    }
+
+    @Override
+    public long getTotalBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public void close()
+    {
+        closed = true;
+    }
+
+    private void assureLoaded()
+    {
+        if (pageIterator != null) {
+            return;
+        }
+        pageIterator = loadPages(tupleDomain);
+    }
+
+    private Iterator<Page> loadPages(TupleDomain<Integer> tupleDomain)
+    {
+        Map<Integer, Domain> domains = tupleDomain.getDomains().get();
+        Domain schemaNameDomain = domains.get(getColumnIndex(tableMetadata, SCHEMA_NAME));
+        Domain tableNameDomain = domains.get(getColumnIndex(tableMetadata, TABLE_NAME));
+
+        String schemaName = schemaNameDomain == null ? null : getStringValue(schemaNameDomain.getSingleValue());
+        String tableName = tableNameDomain == null ? null : getStringValue(tableNameDomain.getSingleValue());
+
+        ImmutableList.Builder<Page> pages = ImmutableList.builder();
+        PageBuilder pageBuilder = new PageBuilder(types);
+
+        Iterator<TableRow> tableRows = dao.getTables(schemaName, tableName).iterator();
+        PeekingIterator<ColumnRow> columnRowIterator = Iterators.peekingIterator(dao.getColumns(schemaName, tableName).iterator());
+
+        while (tableRows.hasNext()) {
+            TableRow tableRow = tableRows.next();
+            while (columnRowIterator.hasNext() && columnRowIterator.peek().getTableId() < tableRow.getTableId()) {
+                columnRowIterator.next();
+            }
+
+            String temporalColumnName = null;
+            ArrayList<String> sortColumnNames = new ArrayList<>();
+            ArrayList<String> bucketColumnNames = new ArrayList<>();
+            while (columnRowIterator.hasNext() && columnRowIterator.peek().getTableId() == tableRow.getTableId()) {
+                ColumnRow columnRow = columnRowIterator.next();
+                if (tableRow.getTemporalColumnId().isPresent() && columnRow.getColumnId() == tableRow.getTemporalColumnId().getAsLong()) {
+                    temporalColumnName = columnRow.getColumnName();
+                }
+                if (columnRow.getSortOrdinalPosition().isPresent()) {
+                    int sortOrdinalPosition = columnRow.getSortOrdinalPosition().getAsInt();
```

Comment:
I don't think this is correct. There aren't any hard guarantees about the value "ordinalPosition", only that it will be an ordering. It should be consistent for the Raptor code, but some external program might have done it differently in the database. (yes, it this should be fixed)

Instead of doing it this way, put them into a `SortedMap` keyed on ordinal position.